airflow/io/store/path.py

```diff
@@ -0,0 +1,666 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import contextlib
+import os
+import typing
+from io import UnsupportedOperation
+from stat import S_ISLNK
+
+from fsspec.utils import stringify_path
+
+from airflow.io.store import ObjectStore, attach
+
+
+def _rewrite_info(info: dict, store: ObjectStore) -> dict:
+    info["name"] = ObjectStoragePath(info["name"], store=store)
+    return info
+
+
+class ObjectStoragePath(os.PathLike):
+    """A path-like object for object storage."""
+
+    sep: typing.ClassVar[str] = "/"
+    root_marker: typing.ClassVar[str] = "/"
+
+    __slots__ = (
+        "_store",
+        "_bucket",
+        "_key",
+        "_conn_id",
+        "_protocol",
+        "_hash",
+    )
+
+    def __init__(self, path, conn_id: str | None = None, store: ObjectStore | None = None):
+        self._conn_id = conn_id
+        self._store = store
+
+        self._hash = None
+
+        self._protocol, self._bucket, self._key = self.split_path(path)
```

Comment:
I wonder if “bucket” is too service-specific; Azure does not use this term, for example. Maybe just the standard scheme, domain, and path is better here.

Reply:
The thing is, that the they are not really standard. `s3://warehouse/` for example would have as a host/netloc "warehouse" which obviously is not what we mean - urlparse("s3://warehouse") = (scheme="s3", netloc="warehouse", path=""). 

`gcs` and `s3` both use bucket and key so I just went where gravity was :-). Besides we deliver our own abstraction here and I am happy to use `bucket` and `key`. They are no **really** exposed anyway.